### PR TITLE
Non zero planning rebased

### PR DESCRIPTION
**Pull request by mahmoud-a-ali**
this PR adds required functions for non-zero segment planning (from non-zero velocity to non-zero velocity trajectory segment).
source files:
- segment_planning.py
- traj_segment.py
- cubic_eq_roots.py

Example: it contains trajectory segments with  different combination of the start and end velocities
- traj_segment_example.py

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
e567258 will 'disappear' from the PR as soon as #6 gets merged, but we thought this would result in a cleaner history (ie: basing @mahmoud-a-ali's PR on-top of #6).

**Comment by mahmoud-a-ali** [comment]
updates:
- update test_file "**test_traj_segment**.py", add more test cases: complex motion cases where velocity direction changes, and unfeasible cases where the test expect an error of a certain type
- add  test_file "**test_traj_segment_middle_points**.py": to test pos/vel/acc/jrk values at each instant the segment_phase changes (these values should be less than limits p_max, v_max, a_max, j_max), besides the main test of the calculated start/end pos/vel 
- add python script "**generate_random_test_cases.py**" to auto-generate random feasible test cases 
- add python script "**generate_execption_test_cases.py**" to auto-generate random unfeasible/illogical test cases
- modify "**plot**.py" function: add p_max limits, plot two horizental dot lines for -p_max, p_max
- modify "**segment_planning**.py" and "**traj_plannig**.py" : add more functions for complex motion planning, raise different error for different unfeasible cases
- add "**plot_traj_segment**.py": to use this function directly to generate plotting script for all auro-generated random/exception cases

**Comment by jonbinney** [comment]
Sorry - I didn't have notifications on for this repo and I completely missed this PR. Will review soon.

File: traj/scripts/traj_segment_example.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    traj.plot.plot_trajectory(plt, position, velocity, acceleration, jerk, n_points=100, v_max=v_max, a_max=a_max, j_max=j_max)
+    plt.show()
+    
+p_max=30
+v_max=3.0
+a_max=4.0
+j_max=10.0   
+
+############### CASE A: v_start = v_end = 0    
+#case A1:  no limit is reached
```
Comment (jonbinney) [inline traj/scripts/traj_segment_example.py:21]: This is an awesome set of tests. Would you mind putting this in the `traj/test` directory and making it a nosetest? Then we can run it automatically as part of the test suite.

File: traj/src/traj/cubic_eq_roots.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+#    print "disc= {}".format(disc)
+    term1 = (b/3.0)
+    if (disc > 0.0): # one root real, two are complex
+#        print "disc > 0"
+        s = r + math.sqrt(disc)
+#        s = ((s < 0) ? -pow(-s, (1.0/3.0)) : pow(s, (1.0/3.0)))
+        if s<0:
+            s= -pow(-s, (1.0/3.0))
+        else:
+            s= pow(s, (1.0/3.0))
```
Comment (jonbinney) [inline traj/src/traj/cubic_eq_roots.py]: Using the `**` operator is a bit more pythonic than `pow()`

File: traj/src/traj/cubic_eq_roots.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    q = (3.0*c - (b*b))/9.0
+    r = -(27.0*d) + b*(9.0*c - 2.0*(b*b))
+    r /= 54.0
+    disc = q*q*q + r*r
+#    print "disc= {}".format(disc)
+    term1 = (b/3.0)
+    if (disc > 0.0): # one root real, two are complex
+#        print "disc > 0"
+        s = r + math.sqrt(disc)
+#        s = ((s < 0) ? -pow(-s, (1.0/3.0)) : pow(s, (1.0/3.0)))
```
Comment (jonbinney) [inline traj/src/traj/cubic_eq_roots.py]: Consider deleting the commented lines. If you have debugging output that you want to keep in the code but not use every time, you can use `rospy.logdebug()`

File: traj/src/traj/cubic_eq_roots.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        rt1 = 0
+        rt2, rt3, n_rts = quad_eq_real_root (a, b, c)
+        return rt1, rt2, rt3, n_rts+1
+    
+    b /= a
+    c /= a
+    d /= a
+#    double disc, q, r, dum1, s, t, term1, r13;
+    q = (3.0*c - (b*b))/9.0
+    r = -(27.0*d) + b*(9.0*c - 2.0*(b*b))
```
Comment (jonbinney) [inline traj/src/traj/cubic_eq_roots.py:28]: Nitpick: would be good to run a formatter on the code to standardize the spacing.

File: traj/src/traj/cubic_eq_roots.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+##### test
+#a=1.2      
+#b=2.3  
+#c=3.4   
+#d=4.5
+##
+#a=2.0 
+#b=0.0 
+#c=0.00099025765642
```
Comment (jonbinney) [inline traj/src/traj/cubic_eq_roots.py]: If this is a test that is useful long term, consider putting it in `traj/test/`

File: traj/scripts/traj_segment_example.py (hunk 1 above)
Comment (mahmoud-a-ali) [reply traj/scripts/traj_segment_example.py:21]: I already added 2 test files with all of these cases in traj/test directory: [test_traj_segment.py](https://github.com/mahmoud-a-ali/trajectory_smoothing/blob/17cfb09e75688e05bf462de2a7d69d229ee429cc/traj/test/test_traj_segment.py#L1), and [test_traj_segment_middle_points.py ](https://github.com/mahmoud-a-ali/trajectory_smoothing/blob/17cfb09e75688e05bf462de2a7d69d229ee429cc/traj/test/test_traj_segment_middle_points.py#L1)

File: traj/src/traj/cubic_eq_roots.py (hunk 2 above)
Comment (mahmoud-a-ali) [reply traj/src/traj/cubic_eq_roots.py]: I will change it. the case that some of the functions I used here, I had implemented in C++ before so I tried to copy paste and make changes!

**Comment by gavanderhoorn** [comment]
@mahmoud-a-ali: we'll have to rebase your work on-top of @jonbinney's.

There are some conflicts already, so we'll need to fix those too.

I can help you with both.

**Comment by mahmoud-a-ali** [comment]
@gavanderhoorn 
- 2 out of the 3 conflecting files: _src/traj/seven_segment.py_, and _test/test_seven_segment.py_ I won't use them anymore, so this 2  conflicts can be fixed easily by using Jon's version of them

- src/traj/plot.py: me and Jon use different footprint for the plot function, I just add one more argument which p_max

File: traj/src/traj/cubic_eq_roots.py (hunk 3 above)
Comment (mahmoud-a-ali) [reply traj/src/traj/cubic_eq_roots.py]: print lines were changed  to 

> rospy.logdebug()

**Comment by mahmoud-a-ali** [comment]
now, this PR can be merged with the current master

File: traj/src/traj/cubic_eq_roots.py (hunk 5 above)
Comment (mahmoud-a-ali) [reply traj/src/traj/cubic_eq_roots.py]: test file is added is traj/test  directory:  [test_min_real_root_cubic_eq.py](https://github.com/mahmoud-a-ali/trajectory_smoothing/blob/5bdd174b3a5b44b6aa79146508957dfbefb6593c/traj/test/test_min_real_root_cubic_eq.py#L1)

**Comment by jonbinney** [comment]
Hmmm... for some reason the diff github is showing looks odd. For example it is showing traj/scripts/demo as an all new file, not a modified file. Tomorrow I'll try to figure out what is going on so we can get this merged.

**Comment by gavanderhoorn** [comment]
I'll try to work with @mahmoud-a-ali to sort the history out a bit.

Rebasing may have not been completely successful.

**Comment by gavanderhoorn** [comment]
~~I've got a cleaner version of this in https://github.com/iron-ox/trajectory_smoothing/compare/master...gavanderhoorn:zero_planning_rebased2.~~

See https://github.com/iron-ox/trajectory_smoothing/pull/8#issuecomment-579720303.

**Comment by jonbinney** [comment]
@gavanderhoorn your branch looks good - it brings the number of changed files down to 19 from the 25 in @mahmoud-a-ali 's branch. Did you mean to add the `.py` to the end of the files in `traj/scripts`? I typically leave executable scripts in ROS packages without an extension and use the shebang to tell the linux how to execute it.

**Comment by gavanderhoorn** [comment]
> Did you mean to add the `.py` to the end of the files in `traj/scripts`?

No. That's probably something that slipped in during one of the rebases. It was a bit complicated to get all the changes to merge cleanly.

I'll take a look.

It's still all @mahmoud-a-ali's work btw. I've only played with the commit history.

**Comment by gavanderhoorn** [comment]
Cleaner version: https://github.com/iron-ox/trajectory_smoothing/compare/master...mahmoud-a-ali:zero_planning_rebased2.

**Comment by mahmoud-a-ali** [comment]
- replaced by PR #10

**Pull request closed without merging**